### PR TITLE
Add CertStore::export_pkcs12

### PR DESCRIPTION
**Pull request by alexcrichton**
This exposes the ability to export a `CertStore` as a PKCS#12 archive. Along
the way this also exports the `memory` constructor to allow easily converting a
`CertContext` with a private key to a PKCS#12 archive. Along the way this
removes the `Memory` type and moves the functions to the `CertStore` type
directly.

### CODE REVIEW

**Comment by alexcrichton** [comment]
The rationale behind this that during local testing when using native-tls I want to create a `CertContext` with winapi functions but then convert that to a `TlsAcceptor` which involves going through the middle-man of a PKCS#12 archive. Feedback welcome though!

**Comment by steffengy** [comment]
from my point of view +1 since supporting basic certificate functionality makes sense.
(required for tests or basic use)

(Yes it would be better to have these sort of APIs in a seperate crate but until it's a bit
more "fleshed out" an extraction into a separate crate doesn't make sense yet,
from my perspective - and one can pretty easily move it to one at a a later stage)

Unifying memory and CertStore is in my oppinion also ok, since I think
it was an abstraction, which didn't really have a use (atleast none i know of)?

@sfackler any objections or feedback from your side?

**Comment by sfackler** [comment]
CertStores are reference counted and I think I was originally unsure if those mutation operations are appropriately thread safe. If they are then this seems fine.

**Comment by steffengy** [comment]
Oh, so it was to prevent people from sharing the underlying CertStore across threads, to prevent
concurrent modifications?

That indeed will be an issue while sharing them is allowed.

There is no documentation regarding thread safety on any of the mutations above,
so I think the sane choice is to assume they aren't.

The [.NET documentation for x509Store](https://msdn.microsoft.com/en-us/library/system.security.cryptography.x509certificates.x509store(v=vs.110).aspx) also hints: 
"Any instance members are not guaranteed to be thread safe."

Technically I also assume they don't take the overhead of ensuring a consistent view
of a certificate store by doing locking or anything similar.

So it probably makes sense to expose a certificate store you can mutate but not share across
threads like before?

Still not sure how that'd work with some "physical certstore" like the current user's cert store,
if it'd be mutated from multiple threads.

**Comment by alexcrichton** [comment]
Good points! Perhaps I could just remove the `Sync` and `Send` from `CertStore`? I think that'd address those points, right?

**Comment by steffengy** [comment]
That'd address those points yes.

In terms of consistency I'm unsure if it's the best solution though,
since some other wrapping types (`CertContext ` and `CtlContext`) 
also implement `Sync` and `Send`.

It also might lead to usability implications, which might've been @sfackler's 
original rationale for making those types `Sync` and `Send`?

In my opinion it still should be cleaner to share a "new" handle to that 
store/context/whatever, when you need access from multiple threads? 
Then it would make sense to remove `Sync` and `Send` on those types.

Still not quite sure how significant the disadvantages in each case would be.

**Comment by sfackler** [comment]
`CertStore` has to be `Sync` and `Send` - otherwise `TlsStream` wouldn't be!

Having a separate builder type and keeping `CertStore` immutable avoids these issues.

**Comment by alexcrichton** [comment]
I'm a little lost now, unfortunately. Could you summarize the changes you'd like to see?

**Comment by sfackler** [comment]
Keep `CertStore` immutable and add back the `Memory` type.

**Comment by alexcrichton** [comment]
And to clarify, `CertStore::add_cert` already exists, but that's ok? Should change?

**Comment by alexcrichton** [comment]
@sfackler thought on [this](https://github.com/steffengy/schannel-rs/pull/15#issuecomment-267796502)? Should those methods be deprecated now if we're worried about reference counting?

In any case I've pushed those changes

**Comment by sfackler** [comment]
Ok, so I think we should deprecate `CertStore::add_cert`, add a `CertStoreBuilder` type that has `memory`, `open_local_machine` and `open_current_user` constructors and does have that method.

**Comment by alexcrichton** [comment]
Perhaps that could happen in a separate PR? I'm already running a little thin as-is and all I wanted to do was update tokio-tls...

I don't really understand the reference counting issues here, and it sounds like they need to be thought through rather than bolting on a solution here?

**Comment by sfackler** [comment]
Works for me.
